### PR TITLE
Bugfix to merging loop in PFEGAlgo (76X)

### PR DESCRIPTION
**Pull request by lgray**
Backport of #12939.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@lgray 
I was hoping for a fix that would have no changes on events that were not failing with an exception before.
The fairly low statistics tests in jenkins show differences in 25.0, 25202.0, 134.911.

You mentioned one in a few billion for the rate of the bug. I'm not sure what was in the denominator or numerator in this rate estimate.

**Comment by lgray** [comment]
That number comes from all the events we've processed before that have not needed this fix. I still find that this assert has shown up now rather strange.

**Comment by lgray** [comment]
The changes in low stats means something is wrong. It should not have an effect on most events. I'll check again.

**Comment by slava77** [comment]
IIRC, the assert happens if maxenergy cluster is not in the same detector as the seed cluster. This may as well add an extra large suppression factor for the rate. So, it may still be consistent.
In this case for 76X I'd rather we have a fix at the point where the symptom is (the assert itself), rather than where the cause is (since if the fix is in the right place, the cause only very rarely leads to the crash).

Thank you for checking.

**Comment by lgray** [comment]
Actually stepping through the logic of the fix it seems there is an issue. Different version incoming.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12940	f6c612d
- code changes look fairly localized/targeted as expected
- jenkins tests pass and comparisons with baseline show no differences 
- previously crashing Run 256676, Event 947703079, LumiSection 673 /Run2015D/JetHT/RAW doesn't crash anymore and generates just one warning 
- additional ~2K events from several workflows tested to gain stats or add some more sanity checks: the warning doesn't show up

File: RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc [hunk 1]
```diff
@@ -1445,6 +1445,18 @@ initializeProtoCands(std::list<PFEGammaAlgo::ProtoEGObject>& egobjs) {
 	 << "Found objects " << std::distance(mergestart,nomerge)
 	 << " to merge by links to the front!" << std::endl;
        for( auto roToMerge = mergestart; roToMerge != nomerge; ++roToMerge) {
+         //bugfix! L.Gray 14 Jan 2016 
+         // -- check that the front is still mergeable!
+         if( thefront.ecalclusters.size() && roToMerge->ecalclusters.size() ) {
```
Comment (davidlange6) [inline RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: size could better be !empty()  [or less size()!=0]. Consider for a follow pr.

**Comment by davidlange6** [comment]
+1

File: RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc (hunk 1 above)
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: "empty()" (begin==end) by itself is cheaper than size.

**Pull request merged**